src/order/lattice.lean

```diff
@@ -305,18 +417,61 @@ end
 
 end semilattice_inf
 
-/-! ### Lattices -/
+/-!
+### Lattices
+-/
 
 /-- A lattice is a join-semilattice which is also a meet-semilattice. -/
 class lattice (α : Type u) extends semilattice_sup α, semilattice_inf α
 
 instance (α) [lattice α] : lattice (order_dual α) :=
 { .. order_dual.semilattice_sup α, .. order_dual.semilattice_inf α }
 
+/--
+A type with a pair of commutative and associative binary operations which satisfy two absorption
+laws relating the two operations has the structure of a lattice.
+
+The partial order is defined so that `a ≤ b` unfolds to `a ⊔ b = b`; cf. `sup_eq_right`.
+-/
+def lattice.mk' {α : Type*} {hs : has_sup α} {hi : has_inf α}
+  (sup_comm : ∀ (a b : α), a ⊔ b = b ⊔ a)
+  (sup_assoc : ∀ (a b c : α), a ⊔ b ⊔ c = a ⊔ (b ⊔ c))
+  (inf_comm : ∀ (a b : α), a ⊓ b = b ⊓ a)
+  (inf_assoc : ∀ (a b c : α), a ⊓ b ⊓ c = a ⊓ (b ⊓ c))
+  (sup_inf_self : ∀ (a b : α), a ⊔ a ⊓ b = a)
+  (inf_sup_self : ∀ (a b : α), a ⊓ (a ⊔ b) = a) : lattice α :=
+have sup_idem : ∀ (b : α), b ⊔ b = b := λ b,
+  calc b ⊔ b = b ⊔ b ⊓ (b ⊔ b) : by rw inf_sup_self
+         ... = b               : by rw sup_inf_self,
+have inf_idem : ∀ (b : α), b ⊓ b = b := λ b,
+  calc b ⊓ b = b ⊓ (b ⊔ b ⊓ b) : by rw sup_inf_self
+         ... = b               : by rw inf_sup_self,
+have sup_eq_iff_inf_eq : ∀ (a b : α), a ⊔ b = b ↔ a ⊓ b = a := λ a b,
+  ⟨λ h, by rw [←h, inf_sup_self],
+   λ h, by rw [←h, sup_comm, inf_comm, sup_inf_self]⟩,
+{ /- Lean doesn't know that the `partial_order` instances from `semilattice_sup` and
+    `semilattice_inf` agree, so we have to provide these fields explicitly. -/
```

Comment:
I think it's worth proving this independently, even though it's more work. That is to say:
```
def semilattice_sup_mk'_partial_order_eq_semilattice_inf_mk'_partial_order {α : Type*}
  [has_sup α]
  [has_inf α]
  (sup_comm : ∀ (a b : α), a ⊔ b = b ⊔ a)
  (sup_assoc : ∀ (a b c : α), a ⊔ b ⊔ c = a ⊔ (b ⊔ c))
  (sup_idem : ∀ (a : α), a ⊔ a = a)
  (inf_comm : ∀ (a b : α), a ⊓ b = b ⊓ a)
  (inf_assoc : ∀ (a b c : α), a ⊓ b ⊓ c = a ⊓ (b ⊓ c))
  (inf_idem : ∀ (a : α), a ⊓ a = a) :
  @semilattice_sup.to_partial_order _ (semilattice_sup.mk' sup_comm sup_assoc sup_idem)
  =
  @semilattice_inf.to_partial_order _ (semilattice_inf.mk' inf_comm inf_assoc inf_idem) := sorry
```

Reply:
These lemmas help with that:
```lean
attribute [ext] has_le

/-- Two preorders are equal if they agree on (≤) -/
@[ext]
lemma preorder.to_has_le_injective {α : Type*} :
  function.injective (@preorder.to_has_le α) :=
λ p q h, begin
  simp_rw [has_le.ext_iff, function.funext_iff, eq_iff_iff] at h,
  casesI p,
  casesI q,
  congr' 1,
  { ext a b, exact h a b },
  { ext a b,
    refine (p_lt_iff_le_not_le a b).trans (iff.trans _ (q_lt_iff_le_not_le a b).symm),
    exact (and_congr (h a b) (not_iff_not_of_iff (h b a))) }
end

/-- Two partial_orders are equal if they agree on their preorder -/
@[ext]
lemma partial_order.to_preorder_injective {α : Type*} ⦃p q : partial_order α⦄ 
  (h : @partial_order.to_preorder _ p = @partial_order.to_preorder _ q) : p = q :=
begin
  casesI p,
  casesI q,
  congr' 1,
  exact congr_arg (@preorder.le _) h,
  exact congr_arg (@preorder.lt _) h,
end
```

Reply:
Ugh, we have these already, it's `preorder.ext` - they're just not tagged `ext` for some reason.

Reply:
Thanks! I can try to add this. Would I be able to use this in `lattice.mk'` or would it be separate?

Reply:
Sure, here's how:
```
def semilattice_sup_mk'_partial_order_eq_semilattice_inf_mk'_partial_order {α : Type*}
  [has_sup α]
  [has_inf α]
  (sup_comm : ∀ (a b : α), a ⊔ b = b ⊔ a)
  (sup_assoc : ∀ (a b c : α), a ⊔ b ⊔ c = a ⊔ (b ⊔ c))
  (sup_idem : ∀ (a : α), a ⊔ a = a)
  (inf_comm : ∀ (a b : α), a ⊓ b = b ⊓ a)
  (inf_assoc : ∀ (a b c : α), a ⊓ b ⊓ c = a ⊓ (b ⊓ c))
  (inf_idem : ∀ (a : α), a ⊓ a = a)
  (sup_inf_self : ∀ (a b : α), a ⊔ a ⊓ b = a)
  (inf_sup_self : ∀ (a b : α), a ⊓ (a ⊔ b) = a) :
  @semilattice_sup.to_partial_order _ (semilattice_sup.mk' sup_comm sup_assoc sup_idem)
  =
  @semilattice_inf.to_partial_order _ (semilattice_inf.mk' inf_comm inf_assoc inf_idem) :=
partial_order.ext $ λ a b, show a ⊔ b = b ↔ a = a ⊓ b, from
  ⟨λ h, by rw [←h, inf_sup_self],
   λ h, by rw [h, sup_comm, inf_comm, sup_inf_self]⟩


/--
A type with a pair of commutative and associative binary operations which satisfy two absorption
laws relating the two operations has the structure of a lattice.

The partial order is defined so that `a ≤ b` unfolds to `a ⊔ b = b`; cf. `sup_eq_right`.
-/

def lattice.mk' {α : Type*} {hs : has_sup α} {hi : has_inf α}
  (sup_comm : ∀ (a b : α), a ⊔ b = b ⊔ a)
  (sup_assoc : ∀ (a b c : α), a ⊔ b ⊔ c = a ⊔ (b ⊔ c))
  (inf_comm : ∀ (a b : α), a ⊓ b = b ⊓ a)
  (inf_assoc : ∀ (a b c : α), a ⊓ b ⊓ c = a ⊓ (b ⊓ c))
  (sup_inf_self : ∀ (a b : α), a ⊔ a ⊓ b = a)
  (inf_sup_self : ∀ (a b : α), a ⊓ (a ⊔ b) = a) : lattice α :=
have sup_idem : ∀ (b : α), b ⊔ b = b := λ b,
  calc b ⊔ b = b ⊔ b ⊓ (b ⊔ b) : by rw inf_sup_self
         ... = b               : by rw sup_inf_self,
have inf_idem : ∀ (b : α), b ⊓ b = b := λ b,
  calc b ⊓ b = b ⊓ (b ⊔ b ⊓ b) : by rw sup_inf_self
         ... = b               : by rw inf_sup_self,
let inf_instance := semilattice_inf.mk' inf_comm inf_assoc inf_idem in
let sup_instance := semilattice_sup.mk' sup_comm sup_assoc sup_idem in
-- here we help lean to see that the two partial orders are equal
let partial_order_inst := @semilattice_sup.to_partial_order _ sup_instance in
have partial_order_eq :
  partial_order_inst = @semilattice_inf.to_partial_order _ inf_instance :=
  semilattice_sup_mk'_partial_order_eq_semilattice_inf_mk'_partial_order _ _ _ _ _ _
    sup_inf_self inf_sup_self,
{ inf_le_left  := λ a b, by { rw partial_order_eq, apply inf_le_left },
  inf_le_right := λ a b, by { rw partial_order_eq, apply inf_le_right },
  le_inf := λ a b c, by { rw partial_order_eq, apply le_inf },
  ..partial_order_inst,
  ..sup_instance,
  ..inf_instance, }
```